examples/event_rock_paper_scissors.jl

```diff
@@ -0,0 +1,221 @@
+        swap_agents!(agent, occupant, model)
+    end
+    return
+end
+
+# The reproduction function is the simplest one.
+
+function reproduce!(agent, model)
+    pos = random_nearby_position(agent, model, 1, pos -> isempty(pos, model))
+    isnothing(pos) && return
+    add_agent!(pos, typeof(agent), model)
+    return
+end
+
+## Defining the propensity and timing of the events
+
+# Besides the actual event action defined as the above functions,
+# there are two more pieces of information necessary:
+# 1) how likely an event is to happen, and
+# 2) how long after the previous event it will happen.
+
+# Now, in the "Gillespie" type of simulations, these two things coincide:
+# The probability for an event is its relative propensity (rate), and the time
+# you have to wait for it to happen is inversely the propensity (rate).
+# When creating an `AgentEvent` (see below), the user has the option to
+# go along this "Gillespie" route, which is the default.
+# However, the user can also have more control by explicitly providing a function
+# that returns the time until an event triggers
+# (by default this function becomes a random sample of an exponential distribution).
+
+# Let's make this concrete. For all events we need to define their propensities.
+# Another way to think of propensities is the relative probability mass
+# for an event to happen.
+# The propensities may be constants or functions of the
+# currently actived agent and the model.
+# Here, the propensities for moving and attacking will be constants,
+
+attack_propensity = 1.0
+movement_propensity = 0.5
+
+# while the propensity for reproduction will be proportional
+# to the population size of the agent
+
+function reproduction_propensity(agent, model)
+    same = count(a -> a isa typeof(agent), allagents(model))
+    return 2*same/nagents(model)
+end
+
+## Creating the `AgentEvent` structures
+
+# Events are registered as an [`AgentEvent`](@ref), then are added into a container,
+# and then given to the [`EventQueueABM`](@ref).
+# The attack and reproduction events affect all agents,
+# and hence we don't need to specify an agent type that this event
+# applies to, leaving the `AbstractAgent` as the default.
+
+attack_event = AgentEvent(action! = attack!, propensity = attack_propensity)
+
+reproduction_event = AgentEvent(action! = reproduce!, propensity = reproduction_propensity)
+
+# The movement event does not apply to rocks however,
+# so we need to specify the agent super type that it applies to,
+# which is `Union{Scissors, Paper}`.
+# Additionally, we would like to change how the timing of the movement events works.
+# We want to change it from an exponential distribution sample to something else.
+# This "something else" is once again an arbitrary Julia function,
+# and for here we will make:
+
+function movement_time(agent, model)
+    ## `agent` is the agent the event will be applied to,
+    ## which we do not use in this function!
+    t = 0.1randn(abmrng(model)) + 1
```

Comment:
small thing, usually we use this convention:

```suggestion
    t = 0.1 * randn(abmrng(model)) + 1
```